game/src/info/person.rs

```diff
@@ -209,26 +209,26 @@ pub fn trips(
         .bg(app.cs.inner_panel)
         .to_geom(ctx, Some(0.3));
         rows.push(
-            Btn::custom(
-                row_btn.clone(),
-                row_btn.color(RewriteColor::Change(app.cs.inner_panel, app.cs.hovering)),
-                hitbox,
-                None,
-            )
-            .build(
-                ctx,
-                format!(
-                    "{} {}",
-                    if open_trips.contains_key(t) {
-                        "hide"
-                    } else {
-                        "show"
-                    },
-                    t
-                ),
-                None,
-            )
-            .margin_above(if idx == 0 { 0 } else { 16 }),
+            ctx.style()
+                .btn_primary_light()
+                .custom_batch(row_btn.clone(), ControlState::Default)
+                .custom_batch(
+                    row_btn.color(RewriteColor::Change(app.cs.inner_panel, app.cs.hovering)),
+                    ControlState::Hovered,
+                )
+                .build_widget(
+                    ctx,
+                    &format!(
```

Comment:
Can you clarify what you're hoping to avoid - is it that you want to avoid sprinkling around `&`?

I guess it's more trauma driven development showing: apps that died the slow death of a billion papercuts from unnecessary string allocation, so as a rule, I prefer &str. In practice the majority of the button builder are static constant &str.

Is it that you don't like the clutter of the syntax or that, in practice, it's a common source of development slowdown due to compiler error?

Reply:
It's partly for consistency; many of the other APIs in widgetry take either `&str` or `String`. I started just accepting `&str`, but there were so many callers needing `format!` that I relaxed some of the calls. It's partly annoying to have get compiler errors (but mitigated by an IDE). The syntax clutter isn't a huge deal to me.

I like the idea of doing the slightly more performant thing by default to avoid having to fix lots of tiny problems later. I realized `Into<String>` is probably the silly choice -- https://doc.rust-lang.org/std/convert/trait.AsRef.html#examples

So what do you think about keeping this `build_widget` API as it is for now, and I'll later go through the whole codebase and switch over to `AsRef<str>` where appropriate?

Reply:
> So what do you think about keeping this build_widget API as it is for now, and I'll later go through the whole codebase and switch over to AsRef<str> where appropriate?

Ah yeah, `AsRef<str>` seems like the best of both worlds, good idea! I'll plan make that change to the builder API now.

Reply:
Hm.. I started down this path, but I don't love having to sprinkle all the `as_ref` around.

```diff
impl Button {
-    fn new(
+    fn new<S: AsRef<str>>(
         ctx: &EventCtx,
         normal: GeomBatch,
         hovered: GeomBatch,
         disabled: GeomBatch,
         hotkey: Option<MultiKey>,
-        action: &str,
+        action: S,
         maybe_tooltip: Option<Text>,
         hitbox: Polygon,
         is_disabled: bool,
@@ -43,16 +43,16 @@ impl Button {
         // dims are based on the hitbox, not the two drawables!
         let bounds = hitbox.get_bounds();
         let dims = ScreenDims::new(bounds.width(), bounds.height());
-        assert!(!action.is_empty());
+        assert!(!action.as_ref().is_empty());
         Button {
-            action: action.to_string(),
+            action: action.as_ref().to_string(),
             draw_normal: ctx.upload(normal),
             draw_hovered: ctx.upload(hovered),
             draw_disabled: ctx.upload(disabled),
             tooltip: if let Some(t) = maybe_tooltip {
                 t
             } else {
-                Text::tooltip(ctx, hotkey.clone(), action)
+                Text::tooltip(ctx, hotkey.clone(), action.as_ref())
             },
             hotkey,
             hitbox,
```


And I'm not even sure what to do about this compiler error:
```
     /// Shorthand method to build a Button wrapped in a Widget
     ///
     /// `action`: The event that will be fired when clicked
-    pub fn build_widget(&self, ctx: &EventCtx, action: &str) -> Widget {
-        Widget::new(Box::new(self.build(ctx, action))).named(action)
+    pub fn build_widget<S: AsRef<str>>(&self, ctx: &EventCtx, action: S) -> Widget {
+        Widget::new(Box::new(self.build(ctx, action))).named(action.as_ref())
     }
 ```

```
   --> widgetry/src/widgets/button.rs:449:62
    |
448 |     pub fn build_widget<S: AsRef<str>>(&self, ctx: &EventCtx, action: S) -> Widget {
    |                                                               ------ move occurs because `action` has type `S`, which does not implement the `Copy` trait
449 |         Widget::new(Box::new(self.build(ctx, action))).named(action.as_ref())
    |                                              ------          ^^^^^^ value borrowed here after move
    |                                              |
    |                                              value moved here
```

I'm actually pretty OK with the current implementation, so I'm not planning on pursuing the AsRef stuff further. I would be happy to have you show me a better way to do this - It's possible I'm doing it wrong.

Reply:
Usually if the arg is used in many places, I do a `let action = action.as_ref();` right at the top. Not sure about that compile error. I can experiment with this refactor later; I don't think it's worth sweating over right now.